File: examples/databases/db-record-indexes.ts
import type { JsBaoClient } from "js-bao-wss-client";

// Index management on a DoDb handle. Add indexes to every field you filter or
// sort on; without one the engine scans all records of that model.
export async function manageIndexes(client: JsBaoClient, databaseId: string) {
  const db = client.databases.connect(databaseId);
  // #region example
  // Register indexes (additive — never drops).
  await db.registerIndex("tasks", "category", "string");
  await db.registerIndex("tasks", "priority", "number");
  await db.registerIndex("appUsers", "email", "string", true); // unique

  // Composite unique constraint across multiple fields.
  await db.registerUniqueConstraint("categories", "name_parentId", ["name", "parentId"]);

  // List and drop.
  const indexes = await db.listIndexes("tasks");
  await db.dropIndex("tasks", "category");

  const constraints = await db.listUniqueConstraints("categories");
  await db.dropUniqueConstraint("categories", "name_parentId");

  // Sync every model passed in the models array at init.
  await db.syncAllIndexes();
  // #endregion example
  return { indexes, constraints };
}

File: examples/databases/db-record-indexes.swift
import JsBaoClient

// Index management on a DoDb handle. Add indexes to every field you filter or
// sort on; without one the engine scans all records of that model.
func manageIndexes(client: JsBaoClient, databaseId: String) async throws {
  let db = client.databases.connect(databaseId: databaseId)
  // #region example
  // Sync the desired index state for one or more models (additive — the
  // server registers only what's missing).
  _ = try await db.syncIndexes(models: [
    DoDbModelSyncState(
      modelName: "tasks",
      indexes: [
        DoDbModelSyncState.Index(fieldName: "category"),
        DoDbModelSyncState.Index(fieldName: "priority", fieldType: "number"),
      ]
    )
  ])

  // Or register indexes manually.
  try await db.registerIndex(modelName: "tasks", fieldName: "category", fieldType: "string")
  try await db.registerIndex(modelName: "tasks", fieldName: "priority", fieldType: "number")
  try await db.registerIndex(modelName: "appUsers", fieldName: "email", fieldType: "string", unique: true)

  // Composite unique constraint across multiple fields.
  try await db.registerUniqueConstraint(modelName: "categories", constraintName: "name_parentId", fields: ["name", "parentId"])

  // List and drop.
  let indexes = try await db.listIndexes(modelName: "tasks")
  try await db.dropIndex(modelName: "tasks", fieldName: "category")

  let constraints = try await db.listUniqueConstraints(modelName: "categories")
  try await db.dropUniqueConstraint(modelName: "categories", constraintName: "name_parentId")
  // #endregion example
  _ = (indexes, constraints)
}
